core/server/worker/src/main/java/alluxio/worker/page/PagedUfsBlockReader.java

```diff
@@ -0,0 +1,299 @@
+   *         content of the requested range
+   */
+  private int fillWithCachedPage(ByteBuffer outBuffer, long offset, long length) {
+    long pageIndex = offset / mPageSize;
+    if (pageIndex != mLastPageIndex) {
+      return 0;
+    }
+    int pageSize = Math.min(mLastPage.remaining(), (int) length);
+    ByteBuffer slice = outBuffer.slice();
+    slice.limit(pageSize);
+    slice.put(mLastPage);
+    mLastPage.rewind();
+    outBuffer.position(outBuffer.position() + pageSize);
+    return pageSize;
+  }
+
+  @Override
+  public long getLength() {
+    return mBlockMeta.getBlockSize();
+  }
+
+  @Override
+  public ReadableByteChannel getChannel() {
+    return getChannel(mInitialOffset);
+  }
+
+  /**
+   * @param offset offset within the block
+   * @return readable channel
+   */
+  public ReadableByteChannel getChannel(long offset) {
+    Preconditions.checkState(!mClosed);
+    return new UfsReadableChannel(offset);
+  }
+
+  /**
+   * @return ufs block read options which this read was created with
+   */
+  public UfsBlockReadOptions getUfsReadOptions() {
+    return mUfsBlockOptions;
+  }
+
+  @Override
+  public int transferTo(ByteBuf buf) throws IOException {
+    Preconditions.checkState(!mClosed);
+    // todo(bowen): eliminate copy
+    ByteBuffer buffer = ByteBuffer.allocateDirect(buf.writableBytes());
+    int bytesRead = transferTo(buffer);
+    buffer.flip();
+    buf.writeBytes(buffer);
+    return bytesRead;
+  }
+
+  int transferTo(ByteBuffer byteBuffer) throws IOException {
+    Preconditions.checkState(!mClosed);
+    int bytesRead = fillWithCachedPage(byteBuffer, mPosition, byteBuffer.remaining());
+    mPosition += bytesRead;
+    try (ReadableByteChannel channel = getChannel(mPosition)) {
+      bytesRead = channel.read(byteBuffer);
+      if (bytesRead < 0) { // eof
+        return bytesRead;
+      }
+      mPosition += bytesRead;
+      return bytesRead;
+    }
+  }
+
+  @Override
+  public boolean isClosed() {
+    return mClosed;
+  }
+
+  @Override
+  public String getLocation() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    mClosed = true;
+  }
+
+  private class UfsReadableChannel implements ReadableByteChannel {
+    private final long mOffset;
+    private volatile InputStream mUfsInStream;
+    private volatile ReadableByteChannel mUfsChannel;
+    private volatile boolean mClosed = false;
+
+    UfsReadableChannel(long offset) {
+      mOffset = offset;
+    }
+
+    @Override
+    public int read(ByteBuffer dst) throws IOException {
+      if (mClosed) {
+        throw new ClosedChannelException();
+      }
+      if (mUfsInStream == null) {
+        synchronized (this) {
+          if (mUfsInStream == null) {
+            UfsManager.UfsClient ufsClient = mUfsManager.get(mUfsBlockOptions.getMountId());
```

Comment:
just an early comment, do you think we can reuse ufsIOManager here to save some duplicate code?

Reply:
+1